File: legacy/python/src/fs_explorer/context_budget.py
"""
Context budget management for LLM calls.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

from google.genai.types import Content, Part

_UNIT_RE = re.compile(r"\[UNIT\s+(\d+)\b", re.IGNORECASE)


@dataclass
class _MessageRecord:
    index: int
    role: str
    original_text: str
    text: str
    bucket: str
    priority: int
    dropped: bool = False


class ContextBudgetManager:
    """Hard-budget context compaction with anchor-centered radial compression."""

    def __init__(
        self,
        *,
        max_input_tokens: int = 12000,
        min_recent_messages: int = 6,
    ) -> None:
        self.max_input_tokens = max_input_tokens
        self.min_recent_messages = min_recent_messages

    def compact_history(
        self,
        history: list[Content],
        *,
        anchor_unit_no: int | None = None,
    ) -> tuple[list[Content], dict[str, int | float | str | bool]]:
        if not history:
            return [], self._empty_stats()

        records = self._build_records(history, anchor_unit_no=anchor_unit_no)
        before_tokens = self._estimate_tokens_records(records)

        self._apply_bucket_caps(
            records,
            caps={
                "current": 3600,
                "ring0": 3000,
                "ring1": 1800,
                "ring2": 900,
                "history": 900,
            },
        )
        if self._estimate_tokens_records(records) > self.max_input_tokens:
            self._apply_bucket_caps(
                records,
                caps={
                    "current": 3000,
                    "ring0": 2200,
                    "ring1": 1200,
                    "ring2": 600,
                    "history": 450,
                },
            )

        if self._estimate_tokens_records(records) > self.max_input_tokens:
            self._drop_oldest_bucket(
                records,
                bucket="history",
                min_keep=self.min_recent_messages,
            )
        if self._estimate_tokens_records(records) > self.max_input_tokens:
            self._drop_oldest_bucket(records, bucket="ring2", min_keep=3)
        if self._estimate_tokens_records(records) > self.max_input_tokens:
            self._drop_oldest_bucket(records, bucket="ring1", min_keep=2)

        if self._estimate_tokens_records(records) > self.max_input_tokens:
            self._apply_bucket_caps(
                records,
                caps={
                    "current": 1200,
                    "ring0": 900,
                    "ring1": 600,
                    "ring2": 350,
                    "history": 220,
                },
            )

        while self._estimate_tokens_records(records) > self.max_input_tokens:
            alive = [record for record in records if not record.dropped]
            if len(alive) <= 1:
                break
            candidates = [
                record
                for record in alive
                if record.bucket != "current" and record.priority < 100
            ]
            if not candidates:
                candidates = alive[:-1]
            if not candidates:
                break
            oldest = min(candidates, key=lambda record: (record.priority, record.index))
            oldest.dropped = True

        self._hard_cap(records)

        output = self._records_to_content(records, history)
        after_tokens = self._estimate_tokens_content(output)
        truncated_messages = sum(
            1 for record in records if not record.dropped and record.text != record.original_text
        )
        dropped_messages = sum(1 for record in records if record.dropped)
        warning = after_tokens > self.max_input_tokens

        stats: dict[str, int | float | str | bool] = {
            "before_tokens": before_tokens,
            "after_tokens": after_tokens,
            "hard_limit_tokens": self.max_input_tokens,
            "truncated_messages": truncated_messages,
            "dropped_messages": dropped_messages,
            "compression_ratio": round(after_tokens / max(before_tokens, 1), 4),
            "overflow_warning": warning,
        }
        if warning:
            stats["warning"] = (
                "Context still exceeds hard budget after compaction; consider "
                "asking for narrower scope."
            )
        return output, stats

    def _hard_cap(self, records: list[_MessageRecord]) -> None:
        """Final deterministic clamp to enforce hard token limits."""
        guard = 0
        while self._estimate_tokens_records(records) > self.max_input_tokens and guard < 200:
            guard += 1
            candidates = [
                record
                for record in records
                if not record.dropped and record.bucket != "current"
            ]
            if not candidates:
                candidates = [record for record in records if not record.dropped]
            if not candidates:
                break
            target = min(candidates, key=lambda record: (record.priority, record.index))
            if len(target.text) > 140:
                target.text = self._compress_text(
                    target.text,
                    max_chars=max(120, int(len(target.text) * 0.65)),
                    bucket=target.bucket,
                )
                continue
            target.dropped = True

    @staticmethod
    def _empty_stats() -> dict[str, int | float | str | bool]:
        return {
            "before_tokens": 0,
            "after_tokens": 0,
            "hard_limit_tokens": 0,
            "truncated_messages": 0,
            "dropped_messages": 0,
            "compression_ratio": 1.0,
            "overflow_warning": False,
        }

    def _build_records(
        self,
        history: list[Content],
        *,
        anchor_unit_no: int | None,
    ) -> list[_MessageRecord]:
        records: list[_MessageRecord] = []
        last_index = len(history) - 1
        for idx, content in enumerate(history):
            text = self._content_text(content)
            bucket = self._classify_bucket(text, anchor_unit_no=anchor_unit_no)
            priority = self._priority_for_bucket(bucket=bucket, idx=idx, last_index=last_index)
            role = getattr(content, "role", "user") or "user"
            records.append(
                _MessageRecord(
                    index=idx,
                    role=str(role),
                    original_text=text,
                    text=text,
                    bucket=bucket,
                    priority=priority,
                )
            )
        return records

    @staticmethod
    def _classify_bucket(text: str, *, anchor_unit_no: int | None) -> str:
        if not text:
            return "history"
        if "Tool result for parse_file" not in text:
            return "history"
        if anchor_unit_no is None:
            return "ring2"

        unit_nos = [int(match.group(1)) for match in _UNIT_RE.finditer(text)]
        if not unit_nos:
            return "ring2"
        distance = min(abs(unit_no - anchor_unit_no) for unit_no in unit_nos)
        if distance == 0:
            return "ring0"
        if distance <= 2:
            return "ring1"
        return "ring2"

    @staticmethod
    def _priority_for_bucket(*, bucket: str, idx: int, last_index: int) -> int:
        if idx == last_index:
            return 100
        if bucket == "ring0":
            return 90
        if bucket == "ring1":
            return 80
        if bucket == "ring2":
            return 70
        return 40 + min(idx, 20)

    def _apply_bucket_caps(
        self,
        records: list[_MessageRecord],
        *,
        caps: dict[str, int],
    ) -> None:
        for record in records:
            if record.dropped:
                continue
            cap = caps.get(record.bucket, caps["history"])
            record.text = self._compress_text(record.text, max_chars=cap, bucket=record.bucket)

    def _drop_oldest_bucket(
        self,
        records: list[_MessageRecord],
        *,
        bucket: str,
        min_keep: int,
    ) -> None:
        while self._estimate_tokens_records(records) > self.max_input_tokens:
            alive = [record for record in records if not record.dropped]
            if len(alive) <= min_keep:
                break
            candidates = [
                record
                for record in records
                if not record.dropped and record.bucket == bucket
            ]
            if not candidates:
                break
            oldest = min(candidates, key=lambda record: record.index)
            oldest.dropped = True

    @staticmethod
    def _compress_text(text: str, *, max_chars: int, bucket: str) -> str:
        if len(text) <= max_chars:
            return text
        if max_chars < 120:
            return text[: max_chars - 20] + "\n...[truncated]..."

        if bucket == "ring2":
            lines = [line.strip() for line in text.splitlines() if line.strip()]
            highlights = [line for line in lines if line.startswith("[UNIT ") or line.startswith("#")]
            highlight_blob = "\n".join(highlights[:8])
            head = text[: max_chars // 3]
            merged = f"{head}\n\n...[ring2 compressed]...\n{highlight_blob}".strip()
            if len(merged) > max_chars:
                return merged[: max_chars - 20] + "\n...[truncated]..."
            return merged

        head = int(max_chars * 0.6)
        tail = max_chars - head - 30
        return f"{text[:head]}\n...[compressed]...\n{text[-tail:]}"

    @staticmethod
    def _content_text(content: Content) -> str:
        parts = getattr(content, "parts", None) or []
        texts: list[str] = []
        for part in parts:
            text = getattr(part, "text", None)
            if text is not None:
                texts.append(str(text))
        return "\n".join(texts).strip()

    def _records_to_content(
        self,
        records: list[_MessageRecord],
        history: list[Content],
    ) -> list[Content]:
        output: list[Content] = []
        by_index = {record.index: record for record in records}
        for idx, original in enumerate(history):
            record = by_index[idx]
            if record.dropped:
                continue
            if record.text == record.original_text:
                output.append(original)
                continue
            output.append(
                Content(
                    role=record.role,
                    parts=[Part.from_text(text=record.text)],
                )
            )
        return output

    @staticmethod
    def _estimate_tokens_text(text: str) -> int:
        # Lightweight, model-agnostic approximation.
        return max(1, (len(text) + 3) // 4)

    def _estimate_tokens_records(self, records: list[_MessageRecord]) -> int:
        return sum(
            self._estimate_tokens_text(record.text)
            for record in records
            if not record.dropped
        )

    def _estimate_tokens_content(self, content: list[Content]) -> int:
        return sum(self._estimate_tokens_text(self._content_text(item)) for item in content)

File: src/agent/context-budget.ts
/*
Reference: legacy/python/src/fs_explorer/context_budget.py
Reference: legacy/python/src/fs_explorer/agent.py
*/

import type { AgentMessage } from "../types/messages.js";

const UNIT_RE = /\[UNIT\s+(\d+)\b/gi;

interface MessageRecord {
  index: number;
  role: string;
  originalText: string;
  text: string;
  bucket: string;
  priority: number;
  dropped: boolean;
}

export interface ContextBudgetStats {
  beforeChars: number;
  afterChars: number;
  hardLimitChars: number;
  droppedMessages: number;
  truncatedMessages: number;
  compressionRatio: number;
  overflowWarning: boolean;
}

export class ContextBudgetManager {
  constructor(
    private readonly config: {
      maxInputChars?: number;
      minRecentMessages?: number;
    } = {},
  ) {}

  compactHistory(
    history: AgentMessage[],
    options: { anchorUnitNo?: number | null } = {},
  ): {
    messages: AgentMessage[];
    stats: ContextBudgetStats;
  } {
    const maxInputChars = this.config.maxInputChars ?? 12_000;
    const minRecentMessages = this.config.minRecentMessages ?? 6;
    if (history.length === 0) {
      return {
        messages: [],
        stats: {
          beforeChars: 0,
          afterChars: 0,
          hardLimitChars: maxInputChars,
          droppedMessages: 0,
          truncatedMessages: 0,
          compressionRatio: 1,
          overflowWarning: false,
        },
      };
    }

    const records = this.buildRecords(history, options.anchorUnitNo ?? null);
    const beforeChars = this.estimateChars(records);

    this.applyBucketCaps(records, {
      current: 3600,
      ring0: 3000,
      ring1: 1800,
      ring2: 900,
      history: 900,
    });
    if (this.estimateChars(records) > maxInputChars) {
      this.applyBucketCaps(records, {
        current: 3000,
        ring0: 2200,
        ring1: 1200,
        ring2: 600,
        history: 450,
      });
    }

    if (this.estimateChars(records) > maxInputChars) {
      this.dropOldestBucket(records, "history", minRecentMessages);
    }
    if (this.estimateChars(records) > maxInputChars) {
      this.dropOldestBucket(records, "ring2", 3);
    }
    if (this.estimateChars(records) > maxInputChars) {
      this.dropOldestBucket(records, "ring1", 2);
    }

    if (this.estimateChars(records) > maxInputChars) {
      this.applyBucketCaps(records, {
        current: 1200,
        ring0: 900,
        ring1: 600,
        ring2: 350,
        history: 220,
      });
    }

    while (this.estimateChars(records) > maxInputChars) {
      const alive = records.filter((record) => !record.dropped);
      if (alive.length <= 1) {
        break;
      }
      let candidates = alive.filter(
        (record) => record.bucket !== "current" && record.priority < 100,
      );
      if (candidates.length === 0) {
        candidates = alive.slice(0, -1);
      }
      if (candidates.length === 0) {
        break;
      }
      candidates.sort((left, right) => left.priority - right.priority || left.index - right.index);
      candidates[0].dropped = true;
    }

    this.hardCap(records, maxInputChars);

    const output = this.recordsToMessages(records, history);
    const afterChars = output.reduce((sum, message) => sum + message.content.length, 0);
    const truncatedMessages = records.filter(
      (record) => !record.dropped && record.text !== record.originalText,
    ).length;
    const droppedMessages = records.filter((record) => record.dropped).length;
    const overflowWarning = afterChars > maxInputChars;

    return {
      messages: output,
      stats: {
        beforeChars,
        afterChars,
        hardLimitChars: maxInputChars,
        droppedMessages,
        truncatedMessages,
        compressionRatio: Number((afterChars / Math.max(beforeChars, 1)).toFixed(4)),
        overflowWarning,
      },
    };
  }

  private buildRecords(history: AgentMessage[], anchorUnitNo: number | null): MessageRecord[] {
    const lastIndex = history.length - 1;
    return history.map((message, index) => {
      const text = message.content;
      const bucket = this.classifyBucket(text, anchorUnitNo);
      return {
        index,
        role: message.role,
        originalText: text,
        text,
        bucket,
        priority: this.priorityForBucket(bucket, index, lastIndex),
        dropped: false,
      };
    });
  }

  private classifyBucket(text: string, anchorUnitNo: number | null): string {
    if (!text) {
      return "history";
    }
    if (!/STRUCTURED CONTEXT PACK|Read receipt:|Document receipt:/i.test(text)) {
      return "history";
    }
    if (anchorUnitNo == null) {
      return "ring2";
    }
    const unitNos = [...text.matchAll(UNIT_RE)].map((match) => Number.parseInt(match[1], 10));
    if (unitNos.length === 0) {
      return "ring2";
    }
    const distance = Math.min(...unitNos.map((unitNo) => Math.abs(unitNo - anchorUnitNo)));
    if (distance === 0) {
      return "ring0";
    }
    if (distance <= 2) {
      return "ring1";
    }
    return "ring2";
  }

  private priorityForBucket(bucket: string, index: number, lastIndex: number): number {
    if (index === lastIndex) {
      return 100;
    }
    if (bucket === "ring0") {
      return 90;
    }
    if (bucket === "ring1") {
      return 80;
    }
    if (bucket === "ring2") {
      return 70;
    }
    return 40 + Math.min(index, 20);
  }

  private applyBucketCaps(records: MessageRecord[], caps: Record<string, number>): void {
    for (const record of records) {
      if (record.dropped) {
        continue;
      }
      const cap = caps[record.bucket] ?? caps.history;
      record.text = this.compressText(record.text, cap, record.bucket);
    }
  }

  private dropOldestBucket(records: MessageRecord[], bucket: string, minKeep: number): void {
    while (this.estimateChars(records) > (this.config.maxInputChars ?? 12_000)) {
      const alive = records.filter((record) => !record.dropped);
      if (alive.length <= minKeep) {
        break;
      }
      const candidates = records.filter(
        (record) => !record.dropped && record.bucket === bucket,
      );
      if (candidates.length === 0) {
        break;
      }
      candidates.sort((left, right) => left.index - right.index);
      candidates[0].dropped = true;
    }
  }

  private hardCap(records: MessageRecord[], maxInputChars: number): void {
    let guard = 0;
    while (this.estimateChars(records) > maxInputChars && guard < 200) {
      guard += 1;
      let candidates = records.filter((record) => !record.dropped && record.bucket !== "current");
      if (candidates.length === 0) {
        candidates = records.filter((record) => !record.dropped);
      }
      if (candidates.length === 0) {
        break;
      }
      candidates.sort((left, right) => left.priority - right.priority || left.index - right.index);
      const target = candidates[0];
      if (target.text.length > 140) {
        target.text = this.compressText(target.text, Math.max(120, Math.floor(target.text.length * 0.65)), target.bucket);
      } else {
        target.dropped = true;
      }
    }
  }

  private compressText(text: string, maxChars: number, bucket: string): string {
    if (text.length <= maxChars) {
      return text;
    }
    if (bucket === "current" || bucket === "ring0") {
      const head = text.slice(0, Math.max(0, Math.floor(maxChars * 0.55))).trimEnd();
      const tail = text
        .slice(Math.max(0, text.length - Math.floor(maxChars * 0.3)))
        .trimStart();
      return `${head}\n...\n${tail}`.slice(0, maxChars);
    }
    return `${text.slice(0, Math.max(0, maxChars - 3)).trimEnd()}...`;
  }

  private estimateChars(records: MessageRecord[]): number {
    return records
      .filter((record) => !record.dropped)
      .reduce((sum, record) => sum + record.text.length, 0);
  }

  private recordsToMessages(records: MessageRecord[], history: AgentMessage[]): AgentMessage[] {
    return records
      .filter((record) => !record.dropped)
      .map((record) => ({
        ...history[record.index],
        content: record.text,
      }));
  }
}
